crates/pathfinder/src/rpc/types.rs

```diff
@@ -785,36 +864,158 @@ pub mod reply {
 
     /// Used in [Block](crate::rpc::types::reply::Block) when the requested scope of
     /// reply is [BlockResponseScope::FullTransactionsAndReceipts](crate::rpc::types::request::BlockResponseScope).
-    ///
-    /// `contract_address` field is available for Deploy and Invoke transactions.
-    /// `entry_point_selector` and `calldata` fields are available only
-    /// for Invoke transactions.
+    #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+    #[serde(untagged)]
+    pub enum TransactionAndReceipt {
+        Declare(DeclareTransactionAndReceipt),
+        Invoke(InvokeTransactionAndReceipt),
+        Deploy(DeployTransactionAndReceipt),
+    }
+
+    impl TransactionAndReceipt {
+        pub fn hash(&self) -> StarknetTransactionHash {
+            match self {
+                TransactionAndReceipt::Declare(declare) => declare.common.txn_hash,
+                TransactionAndReceipt::Invoke(invoke) => invoke.common.txn_hash,
+                TransactionAndReceipt::Deploy(deploy) => deploy.common.txn_hash,
+            }
+        }
+    }
+
     #[serde_as]
-    #[skip_serializing_none]
     #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
-    #[serde(deny_unknown_fields)]
-    pub struct TransactionAndReceipt {
-        pub txn_hash: StarknetTransactionHash,
-        #[serde(default)]
-        pub contract_address: Option<ContractAddress>,
-        #[serde(default)]
-        pub entry_point_selector: Option<EntryPoint>,
-        #[serde(default)]
-        pub calldata: Option<Vec<CallParam>>,
-        #[serde_as(as = "Option<FeeAsHexStr>")]
-        #[serde(default)]
-        pub max_fee: Option<Fee>,
-        #[serde_as(as = "Option<FeeAsHexStr>")]
-        #[serde(default)]
-        pub actual_fee: Option<Fee>,
-        pub status: TransactionStatus,
-        pub status_data: String,
+    pub struct DeclareTransactionAndReceipt {
+        #[serde(flatten)]
+        pub common: CommonTransactionAndReceiptProperties,
+
+        // FIXME: this is a deviation from the OpenRPC spec, which would expect us to
```

Comment:
As I've commented at `DeclareTransaction` above: this is not a deviation anymore.

Reply:
Removed.